core/multigrid/amgx_pgm.cpp

```diff
@@ -0,0 +1,193 @@
+std::unique_ptr<LinOp> amgx_pgm_generate(
+    std::shared_ptr<const Executor> exec,
+    const matrix::Csr<ValueType, IndexType> *source, const size_type num_agg,
+    const Array<IndexType> &agg)
+{
+    auto coarse = matrix::Csr<ValueType, IndexType>::create(
+        exec, dim<2>{num_agg, num_agg}, 0, source->get_strategy());
+    exec->run(amgx_pgm::make_amgx_pgm_generate(source, agg, coarse.get()));
+    return std::move(coarse);
+}
+
+
+}  // namespace
+
+
+template <typename ValueType, typename IndexType>
+void AmgxPgm<ValueType, IndexType>::generate()
+{
+    using matrix_type = matrix::Csr<ValueType, IndexType>;
+    using rmc_value_type = remove_complex<ValueType>;
+    using weight_matrix_type = remove_complex<matrix_type>;
+    auto exec = this->get_executor();
+    const auto num = this->system_matrix_->get_size()[0];
+    Array<IndexType> strongest_neighbor(this->get_executor(), num);
+    Array<IndexType> intermediate_agg(this->get_executor(),
+                                      parameters_.deterministic * num);
+    // Only support csr matrix currently.
+    const matrix_type *amgxpgm_op = nullptr;
+    // Store the csr matrix if needed
+    auto amgxpgm_op_unique_ptr = matrix_type::create(exec);
+    if (!(amgxpgm_op =
+              dynamic_cast<const matrix_type *>(system_matrix_.get()))) {
+        // if original matrix is not csr, converting it to csr.
+        as<ConvertibleTo<matrix_type>>(this->system_matrix_.get())
+            ->convert_to(amgxpgm_op_unique_ptr.get());
+        amgxpgm_op = amgxpgm_op_unique_ptr.get();
+    }
+
+    // Initial agg = -1
+    exec->run(amgx_pgm::make_fill_array(agg_.get_data(), agg_.get_num_elems(),
+                                        -one<IndexType>()));
+    size_type num_unagg{0};
+    size_type num_unagg_prev{0};
+    // TODO: if mtx is a hermitian matrix, weight_mtx = abs(mtx)
+    // compute weight_mtx = (abs(mtx) + abs(mtx'))/2;
+    auto abs_mtx = amgxpgm_op->compute_absolute();
+    // abs_mtx is already real valuetype, so transpose is enough
+    auto weight_mtx = gko::as<weight_matrix_type>(abs_mtx->transpose());
+    auto half_scaler = initialize<matrix::Dense<rmc_value_type>>({0.5}, exec);
+    auto identity = matrix::Identity<rmc_value_type>::create(exec, num);
+    // W = (abs_mtx + transpose(abs_mtx))/2
+    abs_mtx->apply(lend(half_scaler), lend(identity), lend(half_scaler),
+                   lend(weight_mtx));
+    // Extract the diagonal value of matrix
+    auto diag = weight_mtx->extract_diagonal();
+    for (int i = 0; i < parameters_.max_iterations; i++) {
+        // Find the strongest neighbor of each row
+        exec->run(amgx_pgm::make_find_strongest_neighbor(
+            weight_mtx.get(), diag.get(), agg_, strongest_neighbor));
+        // Match edges
+        exec->run(amgx_pgm::make_match_edge(strongest_neighbor, agg_));
+        // Get the num_unagg
+        exec->run(amgx_pgm::make_count_unagg(agg_, &num_unagg));
```

Comment:
nit: You could pass num_unagg as a reference instead of a pointer, which makes for a nicer interface that forbids passing nullptr :)

Reply:
I prefer using the pointer here such that I can easily understand the variable is output.
If do not have strong opinion, I will stay this version.
If it is necessary, we can change it latter